### PR TITLE
Introduce update and delete methods to the AttributeStore

### PR DESCRIPTION
**Pull request by suresh-prakash**

### CODE REVIEW

File: hypertrace-core-graphql-platform/build.gradle.kts [hunk 1]
```diff
@@ -16,7 +16,7 @@ dependencies {
     api("org.hypertrace.core.grpcutils:grpc-client-utils:0.8.2")
     api("org.hypertrace.core.grpcutils:grpc-client-rx-utils:0.8.2")
     api("org.hypertrace.core.grpcutils:grpc-client-rx-utils:0.8.2")
-    api("org.hypertrace.gateway.service:gateway-service-api:0.2.15")
+    api("org.hypertrace.gateway.service:gateway-service-api:0.2.20-SNAPSHOT")
```
Comment (aaron-steinfeld) [inline hypertrace-core-graphql-platform/build.gradle.kts:19]: related?

File: hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/AttributeFilter.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package org.hypertrace.core.graphql.attributes;
+
+import lombok.Builder;
+import lombok.Value;
+import lombok.experimental.Accessors;
+
+@Value
+@Builder
+@Accessors(fluent = true)
+public class AttributeFilter {
```
Comment (aaron-steinfeld) [inline hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/AttributeFilter.java:10]: naming nit: this represents one attribute, not really a generic filter.

File: hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/CachingAttributeStore.java [hunk 3]
```diff
@@ -125,4 +155,22 @@ private NoSuchElementException buildErrorForMissingIdMapping(String scope) {
...
+
+  private AttributeMetadataFilter buildFilter(final AttributeFilter filter) {
+    return AttributeMetadataFilter.newBuilder()
+        .addKey(filter.key())
+        .addScopeString(filter.scope())
+        .build();
+  }
+
+  private List<Update> buildUpdates(final AttributeUpdate update) {
+    final List<Update> updates = new ArrayList<>();
```
Comment (aaron-steinfeld) [inline hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/CachingAttributeStore.java:167]: nit - streaming would read cleaner

File: hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/CachingAttributeStore.java [hunk 4]
```diff
@@ -125,4 +155,22 @@ private NoSuchElementException buildErrorForMissingIdMapping(String scope) {
...
+
+    if (Objects.nonNull(update.displayName())) {
+      updates.add(
+          Update.newBuilder().setDisplayName(Objects.requireNonNull(update.displayName())).build());
```
Comment (aaron-steinfeld) [inline hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/CachingAttributeStore.java:171]: isn't this non null already checked?

**Review by skjindal93 — APPROVED** [review]

File: hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/AttributeUpdate.java [hunk 5]
```diff
@@ -1,13 +1,49 @@
...
+
+  @Value
+  private static class UpdateMapper<T> implements Function<AttributeUpdate, Optional<Update>> {
+    Function<AttributeUpdate, T> valueAccessor;
+    BiFunction<Update.Builder, T, Update.Builder> valueSetter;
+
+    @Override
+    public Optional<Update> apply(final AttributeUpdate attributeUpdate) {
+      final T value = valueAccessor.apply(attributeUpdate);
+      if (Objects.isNull(value)) {
```
Comment (aaron-steinfeld) [inline hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/AttributeUpdate.java:40]: So by making this generic, we lose the ability to have some fields required and others optional. Especially given there's only a single field, this feels like very premature optimization.

File: hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/AttributeUpdate.java [hunk 6]
```diff
@@ -1,13 +1,49 @@
...
+        return Optional.empty();
+      }
+
+      final Update.Builder builder = Update.newBuilder();
+      final Update update = valueSetter.apply(builder, value).build();
```
Comment (aaron-steinfeld) [inline hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/AttributeUpdate.java:45]: someone had fun with this mapper 😆

File: hypertrace-core-graphql-platform/build.gradle.kts (hunk 1 above)
Comment (aaron-steinfeld) [reply hypertrace-core-graphql-platform/build.gradle.kts]: Ah so it's unrelated, you're just piggy backing two changes together? Won't claim I've never done it, but try to discuss both in the pr title/description.

File: hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/AttributeUpdate.java (hunk 6 above)
Comment (suresh-prakash) [reply hypertrace-core-graphql-attribute-store/src/main/java/org/hypertrace/core/graphql/attributes/AttributeUpdate.java:45]: 😆

**Pull request merged**